### PR TITLE
Initial project design

### PR DESCRIPTION
**Pull request by Sammers21**

### CODE REVIEW

**Comment by Sammers21** [comment]
@yegor256 , can you review please?

**Comment by yegor256** [comment]
@Sammers21 try to run `mvn clean install -Pqulice` and see what happens

File: src/main/java/com/yegor256/npm/Meta.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+import java.time.Instant;
+import java.time.ZoneOffset;
+import java.time.ZonedDateTime;
+import java.time.format.DateTimeFormatter;
+import java.util.Set;
+
+/**
+ * The meta.json file
+ */
+public class Meta {
```
Comment (yegor256) [inline src/main/java/com/yegor256/npm/Meta.java]: @Sammers21 why public?

File: src/test/java/com/yegor256/npm/NPMInstallTest.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+import org.junit.Test;
+
+import java.io.File;
+import java.io.IOException;
+
+public class NPMInstallTest {
+
+    @Test
+    public void nomCommandsPublishAndInstallWorks() throws IOException, InterruptedException {
+        int port = 8080;
```
Comment (yegor256) [inline src/test/java/com/yegor256/npm/NPMInstallTest.java]: @Sammers21 not a good idea at all, we should use random TCP port

**Comment by yegor256** [comment]
@Sammers21 see a few comments above. I highly doubt `mvn clean install -Pqulice` will work here... Hava you tried?

File: src/main/java/com/yegor256/npm/Meta.java (hunk 1 above)
Comment (Sammers21) [reply src/main/java/com/yegor256/npm/Meta.java]: The class should be a package private one

File: src/test/java/com/yegor256/npm/NPMInstallTest.java (hunk 2 above)
Comment (Sammers21) [reply src/test/java/com/yegor256/npm/NPMInstallTest.java]: Agree

Comment (Sammers21) [reply src/test/java/com/yegor256/npm/NPMInstallTest.java]: Done

**Comment by Sammers21** [comment]
@yegor256 , can you review once again?

File: src/main/java/com/yegor256/npm/Meta.java [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+    /**
+     * Create meta.json file from uploaded via npm install json file.
+     *
+     * @param published Uploaded json
+     * @param wheretosave Where to store the meta file on disk
+     * @throws IOException on save error
+     */
+    Meta(final JsonObject published, final Path wheretosave)
+        throws IOException {
+        this(Files.write(
```
Comment (yegor256) [inline src/main/java/com/yegor256/npm/Meta.java]: @Sammers21 doesn't look like a good idea, have you read this: https://www.yegor256.com/2015/05/07/ctors-must-be-code-free.html

Comment (yegor256) [reply src/main/java/com/yegor256/npm/Meta.java]: @Sammers21 it seems that you need another class to do this JSON building

File: src/main/java/com/yegor256/npm/Npm.java [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+        );
+        final JsonObject uploaded = reader.readObject();
+        final String metafilename = String.format("%s/meta.json", prefix);
+        final Meta meta;
+        if (this.storage.exists(metafilename)) {
+            this.storage.load(metafilename, metafile);
+            meta = new Meta(metafile);
+            meta.update(uploaded);
+        } else {
+            new Meta(uploaded, metafile);
```
Comment (yegor256) [inline src/main/java/com/yegor256/npm/Npm.java]: @Sammers21 this line looks weird, exactly because of the problem above: ctors must be code-free.

**Comment by yegor256** [comment]
@Sammers21 see my comments above. Also, add this line to `.travis.yml` block `script` and you will see the problem: `gem install xcop`

File: src/main/java/com/yegor256/npm/Meta.java (hunk 3 above)
Comment (Sammers21) [reply src/main/java/com/yegor256/npm/Meta.java]: Just done. Can you re-check if it is ok?

**Comment by Sammers21** [comment]
@yegor256 , can you have another look?

**Comment by yegor256** [comment]
@rultor merge

**Comment by yegor256** [comment]
@Sammers21 thanks!

**Pull request merged**